### PR TITLE
[DRAFT] Frontends as identifiers in Service REST API 

### PR DESCRIPTION
**Pull request by joamaki**
This replaces the use of numerical service identifiers in the service API
and cilium utility with the use of frontend addresses (<ip>:<port>:<protocol>(:<scope>)).

Usage example:
```
    root@838eb2c16285:/home/cilium# cilium service update --frontend 10.0.0.2:80:TCP \
                                      --backends 172.29.0.3:80 --k8s-node-port
    Creating new service "10.0.0.2:80:TCP"
    Added service with 1 backends
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
    
    root@838eb2c16285:/home/cilium# cilium service update --frontend 10.0.1.2:80:TCP \
                                      --backends 172.29.1.3:80,172.29.1.4:80 --k8s-node-port
    Creating new service "10.0.1.2:80:TCP"
    Added service with 2 backends
    
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
    10.0.1.2:80:TCP   NodePort       1 => 172.29.1.3:80 (active)
                                     2 => 172.29.1.4:80 (active)
    
    root@838eb2c16285:/home/cilium# cilium service delete 10.0.1.2:80:TCP
    Service "10.0.1.2:80:TCP" deleted successfully
    
    root@838eb2c16285:/home/cilium# cilium service list
    Frontend          Service Type   Backend
    10.0.0.2:80:TCP   NodePort       1 => 172.29.0.3:80 (active)
```

```release-note
The "cilium service" command now uses frontend address, port and protocol tuple as service identifier instead of the internal numerical identifiers. E.g. instead of "cilium service update --id 123 ..." one now does "cilium service update --frontend 10.0.1.2:80:TCP ...". This change allows Cilium to rearrange the identifiers due to e.g. IP address changes without affecting the users who manipulate the services manually.
```

### CODE REVIEW

File: cilium/cmd/service_update.go [hunk 1]
```diff
@@ -92,37 +66,40 @@ func boolToInt(set bool) int {
 }
 
 func updateService(cmd *cobra.Command, args []string) {
-	warnIdTypeDeprecation()
-
-	id := int64(idU)
-	fa := parseFrontendAddress(frontend)
 	skipFrontendCheck := false
 
+	fa, err := loadbalancer.NewL3n4AddrFromModelID(frontend)
```
Comment (oblazek) [inline cilium/cmd/service_update.go]: This breaks the possibility to update backend state(s)... Usually I do `cilium service update --backends 10.249.97.30:8888 --states maintenance` which means frontend is empty string, but inside the `NewL3n4AddrFromModelID` the frontend format is checked -> i.e. `Error: Cannot update service: invalid service frontend "", expected "<ip>:<port>:<proto>(:i)" e.g. 1.2.3.4:80:TCP (IPv4, external scope) or [2001::1]:443:TCP:i (IPv6, internal scope)`

**Review by oblazek — CHANGES_REQUESTED** [review]

File: cilium/cmd/service_update.go (hunk 1 above)
Comment (joamaki) [reply cilium/cmd/service_update.go]: Thanks for catching this. Will fix this.

**Comment by oblazek** [comment]
this is still wip, right @joamaki ?

File: cilium/cmd/service_update.go (hunk 1 above)
Comment (joamaki) [reply cilium/cmd/service_update.go]: Refactored this now. Could you take another look?

**Comment by oblazek** [comment]
wanted to give you some feedback on the latest version:
```
root@lb-l3-16:~# docker exec -it cilium-l4lb cilium service update --frontend 10.248.11.112:9999:TCP --backends 172.29.0.3:80 --k8s-node-port
>>> {IP:10.248.11.112 Port:9999 Protocol:TCP Scope:external}
Updating existing service "10.248.11.112:9999:TCP"
Error: Cannot add/update service: [PUT /service/{id}][460] putServiceIdInvalidFrontend  the service provided is already registered with ID 64715, please use that ID instead of 0 &{BackendAddresses:[0xc000c40700] Flags:0xc000a23560 FrontendAddress:0xc000c406c0 UpdateServices:false}
```

```
root@lb-l3-16:~# docker exec -it cilium-l4lb cilium service list
Error: Invalid frontend address: unknown L4 protocol
```

**Comment by joamaki** [comment]
Closing this as I need to reopen it under `cilium/cilium` as it's changing `test/l4lb/test.sh` and it can be only tested when under `cilium/cilium`.

**Pull request closed without merging**